File: sorting/quick_sort/quick_sort.dart
void main(List<String> args) {
  List<int> input = [4, 7, 9, 0, 3, 1, 5, 8, 5, 3, 5, 7, 8, 9, 2];
  print(input);
  print("quickSort: ${quickSort(input, 0, input.length - 1)}");
}

quickSort(List<int> elements, int start, int end) {
  if (start < end) {
    int pivot = partition(elements, start, end);
    quickSort(elements, start, pivot - 1);
    quickSort(elements, pivot + 1, end);
  }
  return elements;
}

int partition(List<int> elements, int start, int end) {
  int pivot = elements[end];
  int swapIndex = start - 1;

  for (int i = start; i < end; i++) {
    if (elements[i] <= pivot) {
      swapIndex++;
      swap(elements, swapIndex, i);
    }
  }
  swapIndex++;
  swap(elements, swapIndex, end);

  print("pivot: $pivot");
  print("elements: $elements");

  return swapIndex;
}

void swap(List<int> list, int index1, int index2) {
  int temp = list[index1];
  list[index1] = list[index2];
  list[index2] = temp;
}

File: sorting/quick_sort/quick_sort.js
function partition(elements, start, end) {

    pivot = elements[end]
    swap_index = start - 1;

    console.log(start, end);
    console.log(elements.slice(start, end));
    for (let i = start; i < end; i++) {
        if (elements[i] < pivot) {
            swap_index += 1;

            [[elements[swap_index], [elements[i]]]] = [[elements[i], [elements[swap_index]]]];
        }
    }
    swap_index += 1;
    [[elements[swap_index], [elements[end]]]] = [[elements[end], [elements[swap_index]]]];
    return swap_index;
}

function quick_sort(elements, start, end) {
    if (start < end) {
        pivot = partition(elements, start, end);
        console.log(`pivot: ${pivot}`);
        quick_sort(elements, start, pivot - 1);
        quick_sort(elements, pivot + 1, end);
    }
    return elements;
}

const input = [3, 7, 2, 9, 0, 6];

console.log(input);
console.log(`quick_sort: ${quick_sort(input.slice(), 0, input.length - 1)}`);